File: library/src/main/java/me/shkschneider/skeleton/helper/LogHelper.java
package me.shkschneider.skeleton.helper;

import android.annotation.SuppressLint;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.util.Log;

import me.shkschneider.skeleton.BuildConfig;
import me.shkschneider.skeleton.SkeletonApplication;
import me.shkschneider.skeleton.java.ObjectHelper;
import me.shkschneider.skeleton.java.SkHide;

public class LogHelper {

    protected LogHelper() {
        // Empty
    }

    private static final int VERBOSE = Log.VERBOSE;
    private static final int DEBUG = Log.DEBUG;
    private static final int INFO = Log.INFO;
    private static final int WARN = Log.WARN;
    private static final int ERROR = Log.ERROR;
    private static final int WTF = Log.ASSERT;

    @SkHide
    private static String tag() {
        String tag = SkeletonApplication.TAG;
        if (TextUtils.isEmpty(tag)) {
            tag = BuildConfig.APPLICATION_ID;
        }
        while (tag.length() > 23 && tag.contains(".")) {
            tag = tag.substring(tag.indexOf(".") + 1, tag.length());
        }
        if (tag.length() > 23) {
            tag = tag.substring(tag.length() - 23, tag.length());
        }
        return tag;
    }

    @SuppressLint("LogConditional")
    private static void log(@IntRange(from = VERBOSE, to = WTF) final int level, @Nullable final String msg, @Nullable final Throwable throwable) {
        final StackTraceElement[] elements = new Throwable().getStackTrace();
        String callerClassName = "?";
        String callerMethodName = "?";
        String callerLineNumber = "?";
        if (elements.length >= 3) {
            callerClassName = elements[2].getClassName();
            callerClassName = callerClassName.substring(callerClassName.lastIndexOf('.') + 1);
            if (callerClassName.indexOf("$") > 0) {
                callerClassName = callerClassName.substring(0, callerClassName.indexOf("$"));
            }
            callerMethodName = elements[2].getMethodName();
            callerMethodName = callerMethodName.substring(callerMethodName.lastIndexOf('_') + 1);
            if (callerMethodName.equals("<init>")) {
                callerMethodName = callerClassName;
            }
            callerLineNumber = String.valueOf(elements[2].getLineNumber());
        }
        final String stack = "[" + callerClassName + "." + callerMethodName + "():" + callerLineNumber + "]" + (TextUtils.isEmpty(msg) ? "" : " ");
        switch (level) {
            case VERBOSE:
                Log.v(tag(), stack + msg, throwable);
                break ;
            case DEBUG:
                Log.d(tag(), stack + msg, throwable);
                break ;
            case INFO:
                Log.i(tag(), stack + msg, throwable);
                break ;
            case WARN:
                Log.w(tag(), stack + msg, throwable);
                break ;
            case ERROR:
                Log.e(tag(), stack + msg, throwable);
                break ;
            case WTF:
                Log.wtf(tag(), stack + msg, throwable);
                break ;
            default:
                break ;
        }
    }

    public static void debug(@NonNull final Object object) {
        log(DEBUG, ObjectHelper.jsonify(object), null);
    }

    public static void debug(@NonNull final Object object, @NonNull final Throwable throwable) {
        log(DEBUG, ObjectHelper.jsonify(object), throwable);
    }

    public static void debug(@NonNull final String msg) {
        log(DEBUG, msg, null);
    }

    public static void debug(@NonNull final String msg, @NonNull final Throwable throwable) {
        log(DEBUG, msg, throwable);
    }

    public static void verbose(@NonNull final Object object) {
        log(DEBUG, ObjectHelper.jsonify(object), null);
    }

    public static void verbose(@NonNull final Object object, @NonNull final Throwable throwable) {
        log(DEBUG, ObjectHelper.jsonify(object), throwable);
    }

    public static void verbose(@NonNull final String msg) {
        log(VERBOSE, msg, null);
    }

    public static void verbose(@NonNull final String msg, @NonNull final Throwable throwable) {
        log(VERBOSE, msg, throwable);
    }

    public static void info(@NonNull final Object object) {
        log(DEBUG, ObjectHelper.jsonify(object), null);
    }

    public static void info(@NonNull final Object object, @NonNull final Throwable throwable) {
        log(DEBUG, ObjectHelper.jsonify(object), throwable);
    }

    public static void info(@NonNull final String msg) {
        log(INFO, msg, null);
    }

    public static void info(@NonNull final String msg, @NonNull final Throwable throwable) {
        log(INFO, msg, throwable);
    }

    public static void warning(@NonNull final Object object) {
        log(DEBUG, ObjectHelper.jsonify(object), null);
    }

    public static void warning(@NonNull final Object object, @NonNull final Throwable throwable) {
        log(DEBUG, ObjectHelper.jsonify(object), throwable);
    }

    public static void warning(@NonNull final String msg) {
        log(WARN, msg, null);
    }

    public static void warning(@NonNull final String msg, @NonNull final Throwable throwable) {
        log(WARN, msg, throwable);
    }

    public static void error(@NonNull final Object object) {
        log(DEBUG, ObjectHelper.jsonify(object), null);
    }

    public static void error(@NonNull final Object object, @NonNull final Throwable throwable) {
        log(DEBUG, ObjectHelper.jsonify(object), throwable);
    }

    public static void error(@NonNull final String msg) {
        log(ERROR, msg, null);
    }

    public static void error(@NonNull final String msg, @NonNull final Throwable throwable) {
        log(ERROR, msg, throwable);
    }

    public static void wtf(@NonNull final Object object) {
        log(DEBUG, ObjectHelper.jsonify(object), null);
    }

    public static void wtf(@NonNull final Object object, @NonNull final Throwable throwable) {
        log(DEBUG, ObjectHelper.jsonify(object), throwable);
    }

    public static void wtf(@NonNull final String msg) {
        log(WTF, msg, null);
    }

    public static void wtf(@NonNull final String msg, @NonNull final Throwable throwable) {
        log(WTF, msg, throwable);
    }

    public static void wtf(@NonNull final Throwable throwable) {
        log(WTF, throwable.getMessage(), throwable);
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/helper/LogHelper.kt
package me.shkschneider.skeleton.helper

import android.support.annotation.IntRange
import android.util.Log

import me.shkschneider.skeleton.BuildConfig
import me.shkschneider.skeleton.SkeletonApplication
import me.shkschneider.skeleton.java.ObjectHelper
import me.shkschneider.skeleton.java.SkHide

object LogHelper {

    private const val VERBOSE = Log.VERBOSE
    private const val DEBUG = Log.DEBUG
    private const val INFO = Log.INFO
    private const val WARN = Log.WARN
    private const val ERROR = Log.ERROR
    private const val WTF = Log.ASSERT

    @SkHide
    private fun tag(): String {
        var tag = SkeletonApplication.TAG
        if (tag.isEmpty()) {
            tag = BuildConfig.APPLICATION_ID
        }
        while (tag.length > 23 && tag.contains(".")) {
            tag = tag.substring(tag.indexOf(".") + 1, tag.length)
        }
        if (tag.length > 23) {
            tag = tag.substring(tag.length - 23, tag.length)
        }
        return tag
    }

    private fun log(@IntRange(from = VERBOSE.toLong(), to = WTF.toLong()) level: Int, msg: String?, throwable: Throwable?) {
        val elements = Throwable().stackTrace
        var callerClassName = "?"
        var callerMethodName = "?"
        var callerLineNumber = "?"
        if (elements.size >= 3) {
            callerClassName = elements[2].className
            callerClassName = callerClassName.substring(callerClassName.lastIndexOf('.') + 1)
            if (callerClassName.indexOf("$") > 0) {
                callerClassName = callerClassName.substring(0, callerClassName.indexOf("$"))
            }
            callerMethodName = elements[2].methodName
            callerMethodName = callerMethodName.substring(callerMethodName.lastIndexOf('_') + 1)
            if (callerMethodName == "<init>") {
                callerMethodName = callerClassName
            }
            callerLineNumber = elements[2].lineNumber.toString()
        }
        val stack = "[" + callerClassName + "." + callerMethodName + "():" + callerLineNumber + "]" + if (msg.isNullOrBlank()) "" else " "
        when (level) {
            VERBOSE -> Log.v(tag(), stack + msg, throwable)
            DEBUG -> Log.d(tag(), stack + msg, throwable)
            INFO -> Log.i(tag(), stack + msg, throwable)
            WARN -> Log.w(tag(), stack + msg, throwable)
            ERROR -> Log.e(tag(), stack + msg, throwable)
            WTF -> Log.wtf(tag(), stack + msg, throwable)
        // else -> {}
        }
    }

    fun debug(obj: Any) {
        log(DEBUG, ObjectHelper.jsonify(obj), null)
    }

    fun debug(obj: Any, throwable: Throwable) {
        log(DEBUG, ObjectHelper.jsonify(obj), throwable)
    }

    fun debug(msg: String) {
        log(DEBUG, msg, null)
    }

    fun debug(msg: String, throwable: Throwable) {
        log(DEBUG, msg, throwable)
    }

    fun verbose(obj: Any) {
        log(DEBUG, ObjectHelper.jsonify(obj), null)
    }

    fun verbose(obj: Any, throwable: Throwable) {
        log(DEBUG, ObjectHelper.jsonify(obj), throwable)
    }

    fun verbose(msg: String) {
        log(VERBOSE, msg, null)
    }

    fun verbose(msg: String, throwable: Throwable) {
        log(VERBOSE, msg, throwable)
    }

    fun info(obj: Any) {
        log(DEBUG, ObjectHelper.jsonify(obj), null)
    }

    fun info(obj: Any, throwable: Throwable) {
        log(DEBUG, ObjectHelper.jsonify(obj), throwable)
    }

    fun info(msg: String) {
        log(INFO, msg, null)
    }

    fun info(msg: String, throwable: Throwable) {
        log(INFO, msg, throwable)
    }

    fun warning(obj: Any) {
        log(DEBUG, ObjectHelper.jsonify(obj), null)
    }

    fun warning(obj: Any, throwable: Throwable) {
        log(DEBUG, ObjectHelper.jsonify(obj), throwable)
    }

    fun warning(msg: String) {
        log(WARN, msg, null)
    }

    fun warning(msg: String, throwable: Throwable) {
        log(WARN, msg, throwable)
    }

    fun error(obj: Any) {
        log(DEBUG, ObjectHelper.jsonify(obj), null)
    }

    fun error(obj: Any, throwable: Throwable) {
        log(DEBUG, ObjectHelper.jsonify(obj), throwable)
    }

    fun error(msg: String) {
        log(ERROR, msg, null)
    }

    fun error(msg: String, throwable: Throwable) {
        log(ERROR, msg, throwable)
    }

    fun wtf(throwable: Throwable) { // FIXME
        log(WTF, throwable.message, throwable)
    }

}
